lib/Migration/Version3500Date20231115182612.php

```diff
@@ -0,0 +1,106 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Mail\Migration;
+
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\DB\Types;
+use OCP\IDBConnection;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version3500Date20231115182612 extends SimpleMigrationStep {
+
+	public function __construct(
+		private IDBConnection $connection
+	) {
+	}
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure(): ISchemaWrapper $schemaClosure
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$mailboxesTable = $schema->getTable('mail_mailboxes');
+		if (!$mailboxesTable->hasColumn('name_hash')) {
+			$mailboxesTable->addColumn('name_hash', Types::STRING);
+		}
+
+		return $schema;
+	}
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure(): ISchemaWrapper $schemaClosure
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
+		// Round 1: Hash common mailbox names
+
+		$this->connection->beginTransaction();
+
+		foreach (['INBOX', 'Drafts', 'Sent', 'Trash', 'Junk', 'Spam', 'Archive', 'Archives'] as $name) {
+			$qb = $this->connection->getQueryBuilder();
+			$qb->update('mail_mailboxes')
+				->set('name_hash', $qb->createNamedParameter(md5($name)))
+				->where($qb->expr()->like('name', $qb->createNamedParameter($name), Types::STRING))
+				->executeStatement();
+		}
+
+		$this->connection->commit();
+
+		// Round 2: Hash everything else
+
+		$qb = $this->connection->getQueryBuilder();
+		$qb->select(['id', 'name'])
+			->from('mail_mailboxes')
+			->where($qb->expr()->emptyString('name_hash'));
+		$mailboxes = $qb->executeQuery();
+
+		$updateQb = $this->connection->getQueryBuilder();
+		$updateQb->update('mail_mailboxes')
+			->set('name_hash', $updateQb->createParameter('name_hash'))
+			->where($updateQb->expr()->eq('id', $updateQb->createParameter('id')));
+
+		$this->connection->beginTransaction();
+
+		while (($row = $mailboxes->fetch()) !== false) {
+			$updateQb->setParameter('id', $row['id']);
+			$updateQb->setParameter('name_hash', md5($row['name']));
+			$updateQb->executeStatement();
```

Comment:
nit: for very large installations it could be beneficial if the transaction is broken up every x thousand queries

Reply:
Sure, do you have a number in mind? 

The default size for the innodb log file size https://dev.mysql.com/doc/refman/8.0/en/innodb-parameters.html#sysvar_innodb_log_file_size seems 48mb. I guess that should be enough for a couple of queries ;)

Also interesting: 

https://stackoverflow.com/questions/17555757/transaction-size-limit-in-sql-server
https://stackoverflow.com/questions/709708/maximum-transaction-size-in-postgresql
https://forums.mysql.com/read.php?22,137157,137348#msg-137348

Reply:
for message bulk insertion I did some benchmarks. I think the sweetspot on my local environment was somewhere around 10k entries. this UPDATE here has less data so it's probably capable of more changes in one transaction. 50k?

Reply:
```
I have no name!@7363db6e2186:/var/www/html$ time php occ upgrade
Nextcloud or one of the apps require upgrade - only a limited number of commands are available
You may use your browser or the occ upgrade command to do the upgrade
Setting log level to debug
Turned on maintenance mode
Updating database schema
Updated database
Updating <mail> ...
Restore default tags that are missing

 Done
 5/5 [============================] 100%
Purify and migrate collected mail addresses
 Done
 2/2 [============================] 100%
Insert background jobs for all accounts
 Done
 5/5 [============================] 100%
Updated <mail> to 3.5.0-beta.6
Starting code integrity check...
Finished code integrity check
Update successful
Turned off maintenance mode
Resetting log level

real	0m59.172s
user	0m22.325s
sys	0m8.321s

```

Looks good with 50k batches :+1: